mantidproject/mantid
# Allow loading non-contiguous spectrum number blocks in LoadISISNexus

Fixes #15482

**Context**

LOQ will add an M3 and M4 monitor soon. A first test run on Mantid failed. The reason for this because the way the detectors and monitors are wired up in the following manner:
|----Monitors--||---Detectors---||----Monitors---||----Detectors----|

This cannot be handled currently by the `LoadISISNexus` alogorithm. This issue tries to fix this.

**To test:**
- Please code review
- The unit tests cover several cases  (`LoadISISNexus` is indirectly being tested by a whole bunch of system tests. Since they will use it to load the reference data or test data). The cases of intersperesed monitors which should get handled by this issue cannot be tested directly at this point in time. The reason for this is that the required IDFs are not available or at least should not be in the IDF list of Mantid (not yet). A follow up issue which adds for these cases has been created here:

Please find the required test data here: \olympic\Babylon5\Scratch\Anton\Testing\Framework\PR_15562

**Loading workspace with monitors at the beginning**

The relevant workspace is: LOQ49886.nxs
To test use the instructions below.

**Loading workspace with monitors  at the end**

The relevant workspace is: MER18492.nxs
To test use the instructions below.

**Loading workspace with monitors which are interleaved with the detectors**

The relevant workspace is: LOQ95593.nxs
The relevant IDF is: LOQ_Definition_20160301.xml
You have to add this IDF to your repo your IDF folder. It is important that you remove it once you are done with testing this data set!

To test use the instructions below.

---

_Test instructions:_

Try each of the following instructions below with different settings for 
  `SpectrumMin`, `SpectrumMax`, `SpectrumList`,  `EntryNumber` 
- Press load and select the workspace
- Set:
  - LoadMonitors: Include
    *Confirm that:
  - Detectors and Monitors were loaded together
  - Spectrum numbers match your selection
- Press load and select the workspace
- Set:
  - LoadMonitors: Exclude
- Confirm that:
  - Only detectors are loaded
  - Spectrum numbers match your selection
- Press load and select the workspace
- Set:
  - LoadMonitors: Separate
- Confirm that:
  - Detectors and Monitors are loaded into separate workspaces
  - Spectrum numbers match your selection

[Release notes](https://github.com/mantidproject/mantid/blob/15482_implement_m4_functionality_for_LOQ/docs/source/release/v3.7.0/framework.rst)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
